### ISSUE TITLE
Migrate backlog to github

### ISSUE THREAD

**Issue by epag** [NOAA-OWP/wres#52]
Using this tool I will be migrating the remainder of our backlog to github:

https://github.com/IQSS/redmine2github

Testing will be occurring on a separate repo to ensure this one stays clean and no PII is added by mistake

**Comment by james-d-brown** [comment]
You're starting this on Friday, right? I will probably review one more time.

**Comment by epag** [comment]
> You're starting this on Friday, right? I will probably review one more time.

Yeah, I wont actually start to migrate for real until the end of this week at the earliest. I am just preparing the script/all filtering steps that I will need to take.

**Comment by epag** [comment]
I cant remember who is the vlab admin (I think its James), but can you check that we have rest API integration enabled? We will need that turned on in order to migrate the backlog:

`Most of the time, the API requires authentication. To enable the API-style authentication, you have to check Enable REST API in Administration -> Settings -> API.`

**Comment by epag** [comment]
James or Hank, could one of you check the following above? I don't have an API key on my redmine account so I don't think the above option is selected

**Comment by james-d-brown** [comment]
Hank is the vlab admin, although I am the gerrit admin (but this isn't a gerrit setting).

**Comment by HankHerr-NOAA** [comment]
We use the VLab API to communicate system testing and external service testing results through VLab tickets, so, yes, I think we have it enabled. You should be able to find the credentials to use in the `wres-cron` account, presumably, in order to back those scripts. If you find that some API requests fail, let me know. 

Hank

**Comment by epag** [comment]
> We use the VLab API to communicate system testing and external service testing results through VLab tickets, so, yes, I think we have it enabled. You should be able to find the credentials to use in the `wres-cron` account, presumably, in order to back those scripts. If you find that some API requests fail, let me know.
> 
> Hank

Ah yeah, I forgot about the wres-cron account, let me look into that

**Comment by epag** [comment]
Needed to update the script that I am leveraging some, but I was able to use the wres-cron API key in order to download all of our backlog. Going to create a ticket now to filter out all the tickets that that we flag as not migrating as it looks like the project I am leveraging doesn't have that feature

**Comment by epag** [comment]
Ok, looks like it downloaded the whole backlog. Going to create a script the removes all entries where it contains:

"NOT MIGRATED TO GITHUB"

and doesn't contain (I.E. its been closed) : 
"closed_on": null

**Comment by james-d-brown** [comment]
Should work, providing `closed_on` contains a valid entry when the ticket was rejected, for example.

**Comment by epag** [comment]
I believe it does, but it looks like there is some more work for me to do. It looks like it is pulling the issues from all sub projects on wres. IE, it is getting user support and gui tickets as well. Digging into the redmine api now

**Comment by epag** [comment]
Posting stuff here for myself tomorrow:

Created a python script to remove the unnecessary tickets. Here are the steps:

```
grep -l "WRES User Support" * >> remove.txt
grep -l "WRES Web GUI" * >> remove.txt 
python remover.txt
```

Then I found a way in the download script to specify that I only want open tickets so I will need to redownload them all

**Comment by epag** [comment]
Thats it for today though

**Comment by epag** [comment]
Okay, we are left with 283 tickets so that seems to be a reasonable number based on what I expect. 

Going to attempt to upload these to a fake repo and also going to scrub them for anything

**Comment by epag** [comment]
Looks like ~50 tickets that my script I made tagged as potentially needing redaction. Looking into those

**Comment by epag** [comment]
I was able to port like 6 tickets but now I am getting `requests.exceptions.HTTPError: 403 Client Error: Forbidden for url`

Headers look fine, the token looks good, the request limits seem fine, and it was working initially... Need to do some investigation on what is breaking here. 

This project I am leveraging hasn't been touched since 2015, so there is likely just some assumption that is being broken now

**Comment by epag** [comment]
I may need to look for a different tool or just create our own. It seems like every 3 issues we upload the script crashes for some reason or another (Either creating the issue or uploading a comment). The script is almost 10 years old so it makes sense.

I am going to see if there are any other tools, but if we keep going with this one it could potentially lead to some wonkiness with duplicate issues

**Comment by epag** [comment]
I don't cursory see really any other tools that would do what we need. Ugh, I may just need to migrate all 250 tickets 3 at a time. It may run into some issues linking relevant tickets though

We can discuss in tomorrows meeting

**Comment by epag** [comment]
No luck today figuring things out. There really isnt any rhym or reason as to why the calls are failing. Randomly the script will just fail to make a call creating a ticket or adding a comment and then it just crashes out. It was made as a 1 time migration/hacky so makes sense that it isnt super robust. 

I can retry making the exact call that it fails to make and it will succeed when I run it and there is nothing related to API quotas that we are hitting either.

I am going to attmept to revert python to whatever version it was when this tool released in a VM and see if that is somehow the cause and also increasing the time between requests/timeouts, but otherwise I am going to have to look into updating this tool, making improvements, or coming up with some other script that will allow me to piece meal things.

**Comment by epag** [comment]
Okay good news and bad news. Good news, I believe that I have found the issue with the script, bad news, this script is going to have to run for like a full day without interruption

I finally found this in the github docs:

> Create too much content on GitHub in a short amount of time. In general, no more than 80 content-generating requests per minute and no more than 500 content-generating requests per hour are allowed. Some endpoints have lower content creation limits. Content creation limits include actions taken on the GitHub web interface as well as via the REST API and GraphQL API.

Every single issue created and comment added counts towards that limit. 

And I found this cheeky little comment as to why I wasn't seeing anything about exceeding some rate limit when I was checking that:
> There is not a way to check the status of your secondary rate limit.

Seems like the best way forward is just going to be adding an artificial time inbetween calls...


> no more than 500 content-generating requests per hour are allowed

We will for sure need to make over 500 post requests, so that means im basically limited to making 1 request approximately every 7.5 seconds. which, unsure how many comments we have across those ~100ish tickets, but gonna take a nice long while for this

**Comment by epag** [comment]
But I was basically hitting some ticket that had like 40 comments in it and very quickly would hit these mystery hidden limits

**Comment by james-d-brown** [comment]
Eek. I think we removed a lot of the tickets with, say, 500+ comments, but still...

**Comment by epag** [comment]
Yeah, I am going to add in a sleep and kick this off in my test repo tomorrow morning. Want to make sure it goes smooth before I take it to the main repo (Since you cannot delete issues, just close them)

**Comment by james-d-brown** [comment]
Perhaps we just need to accept this will take a while and perform in batches of, say, 10 tickets as time allows, checking that they completed correctly before moving on...

I don't think we necessarily need to do all this in one big bang. Your call, of course.

**Comment by epag** [comment]
Yeah, that is for sure another approach to consider. Ill see how stable the script is tomorrow with my changes and also I will try out the ticket linking stuff 

(Linking tickets together is the only reason why I think we may need to do it all in one pass as we need the other github tickets to be created in order to reference them)

**Comment by james-d-brown** [comment]
Ah right, gotcha.

**Comment by epag** [comment]
Ugh, I think I have ran into this hangup now. We got much farther (12 issues as opposed to around 3 prior), but we are still hitting the magical "secondary rate limit"

> Some endpoints have lower content creation limits.

I wish they would tell me what those are so I could plan this script accordingly... Gotta love a cryptic wiki pages... 

So basically there is a magical secondary rate, that other than some general rules of thumb (which I am account for in spades), that is blocking this script seemingly randomly. Maybe the comments endpoint is one of those special endpoints. No way to know really. 

I made AROUND 285 post requests in around 40 minutes (could be 295, not sure if the issue creation and initial comment are two posts, but I believe they are one). Of that 285 around 274 were comment posts. 

I **THINK** I know a workaround for if we get paused in our main migration. (Going to go ahead and test my approach now with this break)

However, I think I am going to have to leave my computer running this script overnights and bump the 8 second delay between calls  to something stupid like 15 seconds. I only see limitations on how many calls you can make are per hour, so as long as I can determine the limiting factor I can make things work (Albeit VERY slow)

**Comment by epag** [comment]
![image](https://github.com/user-attachments/assets/ca34ec2d-37be-433b-b3b4-b826dc6a5f46)
This line is such BS though, looks like it should read "Wait 10-20 seconds between requests"

**Comment by epag** [comment]
I calculated things and I believe, that based on what I got throttled at ~280 requests. That rounding that down to 250 an hour, I conservatively get ~14.4

So setting the time between calls to 15 seconds should make it that this script runs without issue. 

In the meantime I am going to scrape the tickets that have migrated to see if there are any keywords we want removed as part of the real migration

**Comment by james-d-brown** [comment]
Just a heads up that I am getting absolutely hammered with e-mails for migrated tickets, currently ticket 149 in a test project. Any way to stop that? I guess it may eed a change in notification settings for a temporary period during migration...

**Comment by epag** [comment]
That is suripising to me, I don't think I forked the project so no idea why you would be getting emailed. I don't think you are a watcher, but maybe you can turn off notifications for the project

https://github.com/epag/wresTicketTesting

**Comment by james-d-brown** [comment]
Yeah, I simply unsubscribed from the ticket 149, but hopefully won't need to do that for each case. I may be some kid of github magic whereby a reference to a github handle in a ticket autosubs you to that ticket. Ticket 149 does contain my gh handle in one of the posted comments...  just a guess. Only received e-mails for 149 so far.

**Comment by epag** [comment]
Im not getting any emails since it is my personal access token I am using I believe so unsure how to help out more. 

ahhh that could maybe be it

**Comment by epag** [comment]
Maybe if you mess with the watch/unwatch setting you can block all notifications 
![image](https://github.com/user-attachments/assets/e27a3b4b-2583-4ff8-bf9c-1201415d29ff)

**Comment by epag** [comment]
If it is what you say then you will probably get emails whenever a ticket mentions you. 

The joys of migrations I guess

**Comment by james-d-brown** [comment]
Yeah. Tbh, that is fine, I was more worried about a notification for every single post in every single migrated ticket, which is what it was starting to look like, but closer inspection revealed that nuance. I'm pretty confident it will only happen when a gh handle is mentioned in a ticket and 149 i probably the only example of that.

**Comment by epag** [comment]
Ive given up trying to find out the secondary rate limits. Fingers crossed at this point my crashing recovery technique works

**Comment by epag** [comment]
Ugh, something annoying I just realized is that the tags were not picked up in the migration (Just labels) so my filtering for "NOT MIGRATED TO GITHUB" did nothing. We still didn't migrate closed items though so there is only about 45 tickets that slipped through.

The fix is simple enough, I just need to add a comment "NOT MIGRATED TO GITHUB" to every **open** ticket and I can filter based on that comment. Apologies for the email spam you all are about to get

**Comment by epag** [comment]
I guess first ill see if I can modify the download script to search for the tag even if it doesn't show in the downloaded result

**Comment by epag** [comment]
Actually looks like I can just export all of the issues via redmine and then ill use that to filter the results, no email spam yay

**Comment by epag** [comment]
Okay, here is an example of how things look with linking to related tickets (Shows both the redmine non-migrated relations and the new github relations if there are any).

A couple things left before the big migration:

- [ ] Fix the random backslash used instead of a forward slash for the redmine links just to make our lives easier
- [ ] Scan through the tickets I have migrated to my private repo for any information I need to scrub outside of the list I already have

![image](https://github.com/user-attachments/assets/b3bed8ce-8c3b-431c-ae0a-5046e855c2ea)

**Comment by epag** [comment]
Let me know if anyone has questions/improvements or wants access to the private repo I created in order to look at the tickets before I do the final migration (Which will take place sometime next week)

**Comment by james-d-brown** [comment]
Looks good. In terms of Redmine related issues, are they actually hyperlinked to VLab? On the one hand, it would be super useful, on the other hand, I'm not sure we can do that. I suppose VLab is easily discoverable, so perhaps it is fine, but I would check with Hank/OWP.

I am also slightly leary about copying across redmine user identifiers exactly, rather than figuratively, although there may not be a great way to mitigate that - most likely it would be a PITA to try.

**Comment by epag** [comment]
They are hyperlinked to VLab, I believe hank said that is fine in an unrelated conversation we had awhile ago, but we can for sure verify with him when he is back next Tuesday or once he is back from Florida. 

I mean assuming I can collect all of the user names it would be simple enough for me to do a sed replacement on those (As I am doing presently using the list of bad words from the github migration). I think the script will just pull anything from the author/comment fields so it doesnt matter if it is a valid username or not

**Comment by james-d-brown** [comment]
Sounds good. If it's OK with OWP, then it's preferable.

Regarding usernames, in that case, I probably would filter them and (for example) replace with first names only (word preceding first stop). Leaking full user ids wouldn't be great, tbh.

**Comment by epag** [comment]
Ill sweep through some tickets and try to find everyones and add that to the replace list

**Comment by epag** [comment]
Can you think of the name of people that may have appeared in these tickets? 

Jesse and Christopher were also on the team yeah? Trying to think of people that would have popped up in our tickets. First names here are enough for me to go through and find their usernames with a grep command

**Comment by james-d-brown** [comment]
Look under wres/settings/members for a complete list, except Jesse, I think, who was removed from the list when he left. All others should be there.

**Comment by james-d-brown** [comment]
In practice, this is a *much* longer list than really needed because 90% of them probably never posted anything.

User support would be trickier as that will have RFC users that are not part of the project, but we aren't migrating those. There could be one or two names that are not project members that slip through, but it would be best effort.

**Comment by epag** [comment]
Ah awesome, thanks, ill take a look there

**Comment by epag** [comment]
Alrighty, I think we are ready for the actual backlog migration (Which I will start tomorrow and ill send out a meeting today marking the start time so people can mute email notifications).

- **Total tickets to migrate:** 214
- **Estimated time to completion** 24-48 hours (depending on how my times I hit the secondary rate limit)
- **How to turn off notifications (RECOMMENDED)**
![image](https://github.com/user-attachments/assets/2e0c4c5c-064e-418f-93b5-e197f528d9ab)

- **Sample ticket**
![image](https://github.com/user-attachments/assets/6abae3c1-6677-4d50-b471-a490b3b0876f)

**Comment by epag** [comment]
Ill leave you all today to bring up anything I should check prior to starting the migration, but otherwise ill start tomorrow before mid day to give everyone a chance to mute their emails. 

ALSO!

Make sure to mark any of the tickets you are working on currently with the current release we are in and/or assigning them to yourself because stuff is about to be burried in a deluge of tickets and that will be the easiest way for you to find tickets going forward.

**Comment by epag** [comment]
Migration has been completed
